ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlAvroDeserializerTest.java

```diff
@@ -53,236 +62,733 @@
+    assertThat(result, is(value));
+  }
+
+  @Test
+  public void shouldCoerceArrayElements() {
+    // Given:
+    givenDeserializerForSchema(SchemaBuilder
+        .array(Schema.OPTIONAL_STRING_SCHEMA)
+        .build()
+    );
+
+    final Map<String, List<?>> validCoercions = ImmutableMap.of(
+        "{\"type\": \"array\", \"items\": \"boolean\"}", ImmutableList.of(true, false),
+        "{\"type\": \"array\", \"items\": \"int\"}", ImmutableList.of(1, 2),
+        "{\"type\": \"array\", \"items\": \"long\"}", ImmutableList.of(1L, 2L),
+        "{\"type\": \"array\", \"items\": \"double\"}", ImmutableList.of(1.1, 2.2),
+        "{\"type\": \"array\", \"items\": \"string\"}", ImmutableList.of("string", "elements!")
+    );
+
+    validCoercions.forEach((avroSchema, value) -> {
+
+      final byte[] bytes = givenAvroSerialized(value, avroSchema);
+
+      final List<String> expected = value.stream()
+          .map(Object::toString)
+          .collect(Collectors.toList());
+
+      // When:
+      final Object result = deserializer.deserialize(SOME_TOPIC, bytes);
+
+      // Then:
+      assertThat(result, is(expected));
+    });
+  }
+
+  @Test
+  public void shouldThrowIfNotAnArray() {
+    // Given:
+    givenDeserializerForSchema(SchemaBuilder
+        .array(Schema.OPTIONAL_STRING_SCHEMA)
+        .build()
+    );
+
+    final byte[] bytes = givenAvroSerialized(true, BOOLEAN_AVRO_SCHEMA);
+
+    // Then:
+    expectedException.expect(SerializationException.class);
+    expectedException.expectCause(hasMessage(containsString(
+        "Cannot deserialize type boolean as type array")));
+
+    // When:
+    deserializer.deserialize(SOME_TOPIC, bytes);
+  }
+
+  @Test
+  public void shouldThrowIfCanNotCoerceArrayElement() {
+    // Given:
+    givenDeserializerForSchema(SchemaBuilder
+        .array(Schema.OPTIONAL_INT32_SCHEMA)
+        .build()
+    );
+
+    final List<?> value = ImmutableList.of("look", "ma,", "an", "array!");
+
+    final byte[] bytes = givenAvroSerialized(value, STRING_ARRAY_AVRO_SCHEMA);
+
+    // Then:
+    expectedException.expect(SerializationException.class);
+    expectedException.expectCause(hasMessage(containsString(
+        "Cannot deserialize type string as type int32")));
+
+    // When:
+    deserializer.deserialize(SOME_TOPIC, bytes);
+  }
+
+  @Test
+  public void shouldDeserializedAvroArrayOfMap() {
+    // Given:
+    givenDeserializerForSchema(SchemaBuilder
+        .array(SchemaBuilder
+            .map(SchemaBuilder.STRING_SCHEMA, Schema.OPTIONAL_INT64_SCHEMA)
+            .optional()
+            .build())
+        .build()
+    );
+
+    final List<?> value = ImmutableList.of(ImmutableMap.of("a", 1L, "b", 2L));
+
+    final byte[] bytes = givenAvroSerialized(value, ARRAY_OF_MAP_AVRO_SCHEMA);
+
+    // When:
+    final Object result = deserializer.deserialize(SOME_TOPIC, bytes);
+
+    // Then:
+    assertThat(result, is(value));
+  }
+
+  @Test
+  public void shouldDeserializedAvroMap() {
```

Comment:
nit: shouldDeserializeAvroMap (ditto for the other names)

Reply:
done in follow up Pr